File: chess_cpp/statecheckers/KingUnderAttackChecker.cpp
//
// Created by Hrach Khachatryan on 12/16/19.
//

#include "KingUnderAttackChecker.h"

void KingUnderAttackChecker::check(Board *board, Spot *start, Spot *end, std::vector<std::string> messages) {
    Spot* kingSpot = board->getKingSpot(!start->getPiece().get()->isWhite());
    std::shared_ptr<Piece> endPiece = end->getPiece();
    end->setPiece(start->getPiece());
    start->setPiece(nullptr);
    KingIsUnderAttackResponse* kingUnderAttack = ((King) kingSpot->getPiece().get()).isKingUnderAttack();
    if(kingUnderAttack->isUnderAttack())
        messages.push_back(kingUnderAttack->getMessage());

    start->setPiece(end->getPiece());
    end->setPiece(endPiece);

}

File: chess_java/src/chess/statecheckers/KingUnderAttackChecker.java
package chess.statecheckers;

import chess.Board;
import chess.Spot;
import chess.beans.KingIsUnderAttackResponse;
import chess.piecies.King;
import chess.piecies.Piece;

import java.util.List;

public class KingUnderAttackChecker extends BaseStateChangeChecker {

    @Override
    public void check(Board board, Spot start, Spot end, List<String> messages) {
        Spot kingSpot = board.getKingSpot(!start.getPiece().isWhite());

        Piece endPiece = end.getPiece();
        end.setPiece(start.getPiece());
        start.setPiece(null);

        KingIsUnderAttackResponse kingUnderAttack = ((King) kingSpot.getPiece()).isKingUnderAttack();
        if(kingUnderAttack.isUnderAttack())
            messages.add(kingUnderAttack.getMessage());

        start.setPiece(end.getPiece());
        end.setPiece(endPiece);

    }
}
